crates/chain/src/keychain/txout_index.rs

```diff
@@ -158,13 +261,30 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     ///
     /// This will panic if a different `descriptor` is introduced to the same `keychain`.
     pub fn add_keychain(&mut self, keychain: K, descriptor: Descriptor<DescriptorPublicKey>) {
-        let old_descriptor = &*self
-            .keychains
+        let descriptor_id = desc_checksum(&descriptor.to_string()).expect("Checksum can't fail");
```

Comment:
I don't think we can assess the implications of having a non-collision-resistant descriptor id. I can imagine applications where this could be a disaster e.g. different descriptor per collaborative custody user where the user can influence the descriptor by choosing their key and grind for a collision with another user. I think we should use a sha256 hash of the string representation of the descriptor without checksum (I don't think there's a binary encoding for descriptors?).

Reply:
+1

Reply:
What do you think about using RIPEMD-160? It's a smaller hash length (which is desirable here) and is already used for bitcoin addresses (so part of rust-bitcoin which is already a dependency).

Reply:
A shorter length is not desirable since that reduces the collision resistance. We either want collision resistance or we don't. If we do we should use the 128-bit security level (SHA256). Note collision resistance is not required for Bitcoin addresses in general -- at least not single user addresses.

Reply:
Yup, makes sense, thanks for pointing this out. 

I'm having troubles hashing a `Descriptor<DescriptorPublicKey>`, has anyone ever tried doing so? It seems to me that `core::hash::Hash` is impl for a descriptor, but not `bitcoin::hashes::Hash` (https://docs.rs/miniscript/latest/miniscript/descriptor/enum.Descriptor.html#impl-Hash-for-Descriptor%3CPk%3E). Am I missing something?

Reply:
use `to_string` and hash the descriptor string as bytes
```rust
    let data = <Vec<u8>>::from(descriptor.to_string().as_bytes());
    let descriptor_id = sha256::Hash::hash(&data);
```